### PR TITLE
compute and summarize gradient noise scale

### PR DESCRIPTION
**Pull request by hnyu**
Add code for computing and summarizing gradient noise scale (GNS). GNS can be used as an indicator  for selecting the mini-batch size (orders of magnitude), and as a rough indicator of the difficulty of a learning task (which might come from the data, the model complexity, or the learning rate).

### CODE REVIEW

File: alf/optimizers/utils.py [hunk 1]
```diff
@@ -31,3 +36,113 @@ def get_opt_arg(p: nn.Parameter, argname: str, default: Any = None):
...
+    def forward(self, loss: torch.Tensor, tensors: alf.nest.NestedTensor):
+        """Given a loss tensor and a nest of tensors, return the estimated GNS.
+
+        Args:
+            loss: a loss tensor *before* taking the mean. Each entry of the tensor
+                represents an individual loss on a single training sample. Ideally,
+                the samples used for computing these losses should be sampled *with*
+                replacement independently. The loss can have a shape of either
+                ``[T,B]`` or ``[B]``. In the first case, we will only take
+                ``[0,B]`` for computing the gns due to sample correlation along
```
Comment (emailweixu) [inline alf/optimizers/utils.py]: why not simply use all?

File: alf/optimizers/utils.py [hunk 2]
```diff
@@ -31,3 +36,113 @@ def get_opt_arg(p: nn.Parameter, argname: str, default: Any = None):
...
+        shuffled_loss = loss[torch.randperm(B)]
+
+        b = max(1, int(B * self._batch_size_ratio))
+        B -= b
+
+        B_norm2 = self._calculate_gradient_norm(shuffled_loss[b:], tensors)
+        b_norm2 = self._calculate_gradient_norm(shuffled_loss[:b], tensors)
+
+        gradient_norm = (B * B_norm2 - b * b_norm2) / (B - b)
+        var_trace = (b_norm2 - B_norm2) / (1. / b - 1. / B)
```
Comment (emailweixu) [inline alf/optimizers/utils.py]: Cause this is quite different from the original formula, do you have some derivation showing that this is what is needed?

File: alf/optimizers/utils.py (hunk 1 above)
Comment (hnyu) [reply alf/optimizers/utils.py]: > why not simply use all?

The assumption of using batch statistics to estimate the true trace and gradient norm is that individual samples of a batch are i.i.d. If T>1, then samples along T are correlated.

File: alf/optimizers/utils.py (hunk 2 above)
Comment (hnyu) [reply alf/optimizers/utils.py]: > Cause this is quite different from the original formula, do you have some derivation showing that this is what is needed?

This is the simplified GNS defined in Appendix A. Compared to the complex one, it doesn't require computing the hessian matrix.

File: alf/optimizers/utils.py (hunk 1 above)
Comment (hnyu) [reply alf/optimizers/utils.py]: > why not simply use all?

Somehow selecting one `t` out of `T` still results in a more stable GNS estimate than using all `t`.

**Comment by hnyu** [comment]
Currently the estimated result is still not stable if no gradient norm clipping is used. The figure shown for sac_cart_pole_conf.py has a norm_clipping=10. Maybe this also implies that we should consider clipping our gradients when training?

**Comment by hnyu** [comment]
Closed as this metric is still not indicative so far.

File: alf/optimizers/utils.py [hunk 3]
```diff
@@ -116,11 +121,10 @@ def forward(self, loss: torch.Tensor, tensors: alf.nest.NestedTensor):
...
                 the samples used for computing these losses should be sampled *with*
                 replacement independently. The loss can have a shape of either
-                ``[T,B]`` or ``[B]``. In the first case, we will only take
-                ``[0,B]`` for computing the gns due to sample correlation along
-                the temporal axis. On-policy algorithms might have noisy estimates
-                if the ``num_parallel_environments`` is small. Similarly, this
-                happens for a small batch size for off-policy algorithms.
+                ``[T,B]`` or ``[B]``. In the first case, we will randomly select
+                a time step ``t``. The reason is that if the ``T`` steps' gradients
+                are consider together, it will violate the i.i.d. assumption of
```
Comment (emailweixu) [inline alf/optimizers/utils.py]: Treating each sequence as a sample, the samples of a batch are still IID.
And using a single t does not work for RNN.

Comment (hnyu) [reply alf/optimizers/utils.py]: > Treating each sequence as a sample, the samples of a batch are still IID. And using a single t does not work for RNN.

I think as long as the subsequence is randomly sampled with a random start position, using a single t still works (produces meaning statistics for estimating GNS). But so far it turns out neither is very stable. 

Note that each loss is already computed before taking the gradients, so it has considered the temporal dependency. Taking one t won't change this.

Comment (hnyu) [reply alf/optimizers/utils.py:126]: > Treating each sequence as a sample, the samples of a batch are still IID. And using a single t does not work for RNN.

In fact, there are multiple losses computed along a sequence, each loss for a time step. Even though sequences are iid with each other, the losses are not from iid samples?

**Comment by hnyu** [comment]
Further experiments show that `sac_cart_pole_conf.py` is a very special case. Probably related to its behavior of directly getting action distributions from Q values, which makes the gradient of its critic networks unstable (not sure how to address this issue yet). For other continuous action cases, SAC is fine with this GNS estimator.

image

The last "skills learning" task also uses SAC.

**Review by emailweixu — APPROVED** [review]
